### PR TITLE
Implement proposed API for `proc_macro_span`

### PR DESCRIPTION
**Pull request by jhpratt**
As proposed in [#54725 (comment)](https://github.com/rust-lang/rust/issues/54725#issuecomment-1546918161). I have omitted the byte-level API as it's already available as [`Span::byte_range`](https://doc.rust-lang.org/nightly/proc_macro/struct.Span.html#method.byte_range).

@rustbot label +A-proc-macros

r? @m-ou-se

### CODE REVIEW

File: tests/ui/macros/same-sequence-span.stderr
Comment (jhpratt) [inline tests/ui/macros/same-sequence-span.stderr:1]: This is something that needs addressing. I presume it has something to do with the ordering of spans, but beyond that I'm not sure what.

**Review by m-ou-se — APPROVED** [review]
This looks fine to me, but I don't know what's up with `tests/ui/macros/same-sequence-span.stderr`.

File: tests/ui/macros/same-sequence-span.stderr
Comment (m-ou-se) [reply tests/ui/macros/same-sequence-span.stderr:1]: @estebank Do you perhaps know what happened here? Both the old and the new diagnostic output seems a bit off here. Not sure what span it is trying to point at. (And also not sure why it changed in this PR.)

Comment (estebank) [reply tests/ui/macros/same-sequence-span.stderr:1]: It seems like the underlying `Span` coming out of the macro is absolutely malformed, very likely the `end` comes before the start, but this is the first time I see this behavior in the test suite (things might have changed in the handling in the last six months). I'd need to dig deeper.

Comment (m-ou-se) [reply tests/ui/macros/same-sequence-span.stderr:1]: I beleive @dtolnay thought might be the result of error deduplication, when I asked him about it.

Comment (jhpratt) [reply tests/ui/macros/same-sequence-span.stderr:1]: If the erroneous output is at most tangentially related to this PR, should it be merged and a separate issue created for the bug?

Comment (m-ou-se) [reply tests/ui/macros/same-sequence-span.stderr:1]: > should it be merged and a separate issue created for the bug?

I think that's the best way forward. Can you open that issue?

File: library/proc_macro/src/lib.rs [hunk 1]
```diff
@@ -494,28 +493,32 @@ impl Span {
...
-    }
-
-    /// Gets the ending line/column in the source file for this span.
-    #[unstable(feature = "proc_macro_span", issue = "54725")]
-    pub fn end(&self) -> LineColumn {
-        self.0.end().add_1_to_column()
-    }
-
     /// Creates an empty span pointing to directly before this span.
     #[unstable(feature = "proc_macro_span_shrink", issue = "87552")]
```
Comment (m-ou-se) [inline library/proc_macro/src/lib.rs:497]: Let's merge the features:
```suggestion
    #[unstable(feature = "proc_macro_span", issue = "54725")]
```

File: library/proc_macro/src/lib.rs [hunk 2]
```diff
@@ -494,28 +493,32 @@ impl Span {
...
-    pub fn before(&self) -> Span {
-        Span(self.0.before())
+    pub fn start(&self) -> Span {
+        Span(self.0.start())
     }
 
     /// Creates an empty span pointing to directly after this span.
     #[unstable(feature = "proc_macro_span_shrink", issue = "87552")]
```
Comment (m-ou-se) [inline library/proc_macro/src/lib.rs:503]: Let's merge the features:
```suggestion
    #[unstable(feature = "proc_macro_span", issue = "54725")]
```

**Review by m-ou-se — APPROVED** [review]

**Comment by m-ou-se** [comment]
Implementation looks fine to me. r=me with that issue filed and conflict resolved.

**Comment by jhpratt** [comment]
cc @dtolnay — looks like merging `proc_macro_span` and `proc_macro_span_shrink` cause `proc_macro2` to fail to compile again.

**Comment by jhpratt** [comment]
Not sure what to make of that error.

**Pull request merged**